processing/src/main/java/io/druid/segment/virtual/SingleStringInputCachingExpressionColumnValueSelector.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.virtual;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import io.druid.java.util.common.ISE;
+import io.druid.math.expr.Expr;
+import io.druid.math.expr.ExprEval;
+import io.druid.math.expr.Parser;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.ColumnValueSelector;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.data.IndexedInts;
+import it.unimi.dsi.fastutil.ints.Int2ObjectLinkedOpenHashMap;
+
+import javax.annotation.Nullable;
+
+/**
+ * Like {@link ExpressionColumnValueSelector}, but caches results for the first CACHE_SIZE dictionary IDs of
+ * a string column.
+ */
+public class SingleStringInputCachingExpressionColumnValueSelector implements ColumnValueSelector<ExprEval>
+{
+  private static final int CACHE_SIZE = 1000;
+
+  private final DimensionSelector selector;
+  private final Expr expression;
+  private final Expr.ObjectBinding bindings;
+  private final ExprEval[] arrayEvalCache;
+  private final LruEvalCache lruEvalCache;
+
+  public SingleStringInputCachingExpressionColumnValueSelector(
+      final DimensionSelector selector,
+      final Expr expression
+  )
+  {
+    // Verify expression has just one binding.
+    if (Parser.findRequiredBindings(expression).size() != 1) {
+      throw new ISE("WTF?! Expected expression with just one binding");
+    }
+
+    this.selector = Preconditions.checkNotNull(selector, "selector");
+    this.expression = Preconditions.checkNotNull(expression, "expression");
+
+    if (selector.getValueCardinality() == DimensionSelector.CARDINALITY_UNKNOWN) {
+      throw new ISE("Selector must have a dictionary");
+    }
+
+    final Supplier<Object> inputSupplier = ExpressionSelectors.supplierFromDimensionSelector(selector);
+    this.bindings = name -> inputSupplier.get();
+
+    if (selector.getValueCardinality() <= CACHE_SIZE) {
```

Comment:
Looks like it doesn't consider -1 which is "unknown" cardinality

Reply:
There is a check for this a couple of lines up… cardinality >= 0 is a precondition for using this optimized implementation. I moved the check to be closer to this block so they are more visually connected. I also added a note to the javadoc about the precondition.